docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,202 @@
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce" / "Sync"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+```
+
+Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. "Sync" mode will include reapplying the resources on resource hash change or periodically following an interval.
+Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.
+
+*Sample Secret Format*
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name:calico-addon
+type: Opaque
+stringData:
+  value: |-
+    kind: ConfigMap
+    apiVersion: v1
+    metadata:
+     name: calico-conf
+```
+
+The resources in ClusterResourceSet will be applied to matching clusters. 
+A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: capi-crs-my-awesome-cluster
+  namespace: cluster-api-system
+data:
+  <ClusterResourceSet-name1>:   
+    <secret-name1>:
+      hash: <>
+      status: success
+      error: ""
+      timestamp: "2020-04-05T08:24:17Z"
+    <configmap-name1>:
+      hash: <>
+      status: failed
+      error: "some error"
+      timestamp: "2020-05-05T08:24:17Z"
+  <ClusterResourceSet-name2>:  
+    <secret-name2>:
+      hash: <>
+      status: success
+      error: ""
+      timestamp: "2020-04-05T08:24:17Z"
+  Status: InProgress/Completed
+```
+Status will be `Completed` when all matching ClusterResourceSet reconciles are completed for that cluster. In case of new resource addition to a matching ClusterResourceSet, Status becomes `InProgress`
+Also, the errors / overall progress will be tracked in the ClusterResourceSet’s status.
```

Comment:
In what field(s)?

Reply:
Does it make sense to use Status field to log permanent events like this:
```
status:
 LastUpdated:  "2020-05-05T08:24:17Z"
 StatusMessages: {
                    {Message: "Added calico-addon resource", Type: "Update"},
                    {Message: "Failed applying to x cluster", Type: "Error"}, 
                }
```

Reply:
We could make use of `Conditions` in the future as well

Reply:
Is there a locking problem for this config map, because of different ClusterResourceSet reconciling into the same object

Reply:
@fabriziopandini 
Writing to the same ClusterResourceSetStatus may be a problem as multiple ClusterResourceSet controller threads may be reaching to the same ClusterResourceSetStatus (This is the new CRD that keeps track of which resources are applied to a cluster, recently added). 
This issue can be overcomed by making ClusterResourceSetStatus clusterresourcesets a map. cc @detiber

Reply:
a map would be potentially easier to handle modifying a single resource from multiple writers